tmt/steps/report/polarion.py

```diff
@@ -213,93 +223,120 @@ def go(self, *, logger: Optional[tmt.log.Logger] = None) -> None:
             # as with steps on cmdline
             if param:
-                properties[f"polarion-custom-{tr_field.replace('_', '')}"] = param
+                testsuites_properties[f"polarion-custom-{tr_field.replace('_', '')}"] = param
+
         if use_facts:
-            properties['polarion-custom-hostname'] = \
-                self.step.plan.provision.guests()[0].primary_address
-            properties['polarion-custom-arch'] = self.step.plan.provision.guests()[0].facts.arch
+            guests = self.step.plan.provision.guests()
+            try:
+                testsuites_properties['polarion-custom-hostname'] = guests[0].primary_address
+                testsuites_properties['polarion-custom-arch'] = guests[0].facts.arch
+            except IndexError as e:
+                raise tmt.utils.ReportError(
+                    f'Failed to get the Polarion facts from `step.plan.provision.guests()`: {e}')
+
         if template:
-            properties['polarion-testrun-template-id'] = template
+            testsuites_properties['polarion-testrun-template-id'] = template
+
         logs = os.getenv('TMT_REPORT_ARTIFACTS_URL')
-        if logs and 'polarion-custom-logs' not in properties:
-            properties['polarion-custom-logs'] = logs
-        testsuites_properties = ElementTree.SubElement(xml_tree, 'properties')
-        for name, value in properties.items():
-            ElementTree.SubElement(testsuites_properties, 'property', attrib={
-                'name': name, 'value': str(value)})
-
-        testsuite = xml_tree.find('testsuite')
-        project_span_ids = xml_tree.find(
-            '*property[@name="polarion-project-span-ids"]')
-
-        for result in self.step.plan.execute.results():
+        if logs and 'polarion-custom-logs' not in testsuites_properties:
+            testsuites_properties['polarion-custom-logs'] = logs
+
+        project_span_ids: list[str] = []
+
+        results_context = ResultsContext(self.step.plan.execute.results())
+
+        for result in results_context:
             if not result.ids or not any(result.ids.values()):
                 self.warn(
                     f"Test Case '{result.name}' is not exported to Polarion, "
                     "please run 'tmt tests export --how polarion' on it.")
                 continue
+
             work_item_id, test_project_id = find_polarion_case_ids(result.ids)
 
             if test_project_id is None:
                 self.warn(f"Test case '{result.name}' missing or not found in Polarion.")
                 continue
 
-            assert work_item_id is not None
-            assert project_span_ids is not None
+            if work_item_id is None:
+                raise tmt.utils.ReportError(
+                    f"Work item ID missing or not found in Polarion (Test case: '{result.name}', "
+                    "Test project ID: '{test_project_id}).")
```

Comment:
Is really `work_item_id` unreachable? This would mean if `work_item_id==None` then always `test_project_id==None`.

Is this always true? `if test_project_id is None -> work_item_id is None`

In that case, we could just change the condition to `if None in [work_item_id, test_project_id]: ...` or just add `or work_item_id is None` as you recommended.

Reply:
yeah, either it found the item and filled both values or it didn't and neither of the values are set...assuming everything went fine in creating the instance in pylero

yeah, whichever condition is fine with me

Reply:
Ok, changed. My construction of condition does not pass the typing checks. I think you can mark this resolved.